design/httpproxy-status-conditions.md

```diff
@@ -0,0 +1,279 @@
+# Status Conditions for HTTPProxy
+
+Status: Draft
+
+## Abstract
+
+This document describes a design and schema for adding Conditions to HTTPProxy’s `status` section. The purpose of these Conditions is to allow a HTTPProxy user to have an accurate view of what the HTTPProxy is configuring Contour to do.
+
+## Background
+
+HTTPProxy historically has had a very basic `status` section, which indicated whether it was “valid” or “invalid”, in the Validity field, and a reason, in the Description field.
+Because HTTPProxy models a complex domain, there are many ways in which a configuration can be invalid, and it’s helpful to the user to be able to expose more than a single one at once.
+
+In addition, it’s possible for a HTTPProxy or set of HTTPProxies to produce a partially-valid configuration, where Contour will configure Envoy with *part* of the desired config, but not all. The status should be able to represent this.
+
+The Kubernetes API standard way to represent states in an extensible way is via a `conditions` stanza, defined in the [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md), but with additional conversation ongoing at time of writing in PR [#4521](https://github.com/kubernetes/community/pull/4521). As a result of this, we've tried to steer a middle path here, and add only one top-level condition, with a similar behavior to existing conditions (that is, Ready). In addition, [KEP-1623](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions) has been moved to implementable, so what Conditions we do add will conform to that schema.
+
+
+## Goals
+
+- To define a standard Condition schema for HTTPProxy and its sibling CRDs (currently TLSDelegation, but should be applicable to any new CRDs also).
+- To add a place in the CRDs for the Conditions to live.
+
+## Non-goals
+Add support for some method of exposing the inclusion graph in the HTTPProxy’s status. That is a separate issue (#xxx)
+
+## High-level design
+
+We're going to do the following things:
+- add a `conditions` section to HTTPProxy `status`. We're doing this to allow for additional extensibility in the standard way.
+- Add a `Valid` condition as the standard HTTPProxy condition. We've chosen a `Valid` condition here rather than a more traditional `Ready` because `Ready` implies that we're sure that the Envoy config has been accepted - which we don't have the mechanisms to ensure yet. We may add a `Ready` condition at a later date.
+- The `Valid` condition's `Reason` field and the `validity` field will have the same value, as will the `Valid` condition's message field and the `description` field. That is, we will keep the functionality of the `validity` and `description` fields, but also duplicate it in the `Valid` Condition.
```

Comment:
There are a few ideas I'm working towards here:
- keep the existing fields with their existing behavior. This is unavoidable, because of v1 API guarantees.
- Add a Condition so that people can use things that expect Conditions (like `kubectl wait --for condition=`), and to have a place to put other conditions in the future.
- Make it so that there are multiple ways for errors to be sourced if there are multiple errors, on the assumption that a HTTPProxy user shouldn't need lots of passes to fix their object if it's broken. This is a current problem with the single reason setup.
- Try to keep the polarity of the top-level conditions consistent. Having a `Validated` condtion with normal-true polarity and `ValidationError` and `ValidationWarning` conditions with abnormal-true polarities seems also not ideal.

The last two things are the reasons why I thought that sub-conditions would be better.

I'll grant that changing the condition schema is a bigger deal than I initially thought though.